Panic on non-utf-8 status messaage

Please answer these questions before submitting your issue.

### What version of gRPC are you using?
grpc-go 1.11

### What version of Go are you using (`go version`)?
go version go1.8.4 linux/amd64

### What operating system (Linux, Windows, …) and version?
Linux

### What did you do?
Server returns status with message which is invalid utf-8 sequence first without details and after that with some details.

### What did you expect to see?
Consistent behavior, for example sanitizing message to ensure that it's valid utf-8.

### What did you see instead?
Returning status with non-utf-8 message but without details is silently encoded by server and is sent to client. Resulting `status-message` http2 field is not valid escaped utf-8 sequence in violation of http2 protocol spec for grpc.

Returning status with non-utf-8 message and with at least one detail causes panic on a server.

I've used `string([]byte{0xff, 0xfe, 0xfd})` as a message, issue can be reproduced by altering `SayHello` method of greeter server from grpc-go examples: https://gist.github.com/euroelessar/dde1d70cf8e8fe8dd27a81d6b2b9025f

Here is a panic stack trace caused by non-utf-8 message:
```
panic: proto: invalid UTF-8 string

goroutine 10 [running]:
google.golang.org/grpc/transport.(*http2Server).WriteStatus(0xc420172000, 0xc4201520f0, 0xc4201be018, 0xc420184001, 0xc420182070)
	/Users/elessar/go/src/google.golang.org/grpc/transport/http2_server.go:763 +0x141c
google.golang.org/grpc.(*Server).processUnaryRPC(0xc420154000, 0x1742a60, 0xc420172000, 0xc4201520f0, 0xc42007dc80, 0x17752a0, 0x0, 0x0, 0x0)
	/Users/elessar/go/src/google.golang.org/grpc/server.go:1023 +0xe65
google.golang.org/grpc.(*Server).handleStream(0xc420154000, 0x1742a60, 0xc420172000, 0xc4201520f0, 0x0)
	/Users/elessar/go/src/google.golang.org/grpc/server.go:1249 +0x1528
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc420018750, 0xc420154000, 0x1742a60, 0xc420172000, 0xc4201520f0)
	/Users/elessar/go/src/google.golang.org/grpc/server.go:680 +0x9f
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/elessar/go/src/google.golang.org/grpc/server.go:678 +0xa1
```

> IIUC, encoded messages should all be ASCII. It's not clear to me how this will result in violation of http2 protocol. An example would help here.

[gRPC HTTP2 protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md ) is a bit more precise:
> The value portion of **Status-Message** is conceptually a Unicode string description of the error, physically encoded as UTF-8 followed by percent-encoding.

But grpc-go doesn't perform any validation of message returned by application handler and therefore sends ASCII encoded string followed by percent encoding which is not always a valid UTF-8 string.

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -6170,7 +6170,9 @@ func TestStatusInvalidUTF8Message(t *testing.T) {
 	}
 }
 
-// Service handler returns status with invalid utf8 details.
+// Service handler returns status with details and invalid utf8 message. Proto
+// will fail to marshal the status because of the invalid utf8 message. Details
+// will be dropped when sending.
 func TestStatusInvalidUTF8Details(t *testing.T) {
 	defer leakcheck.Check(t)
 
@@ -6197,8 +6199,14 @@ func TestStatusInvalidUTF8Details(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
 	defer cancel()
 
-	if _, err := ss.client.EmptyCall(ctx, &testpb.Empty{}); status.Convert(err).Message() != wantMsg {
-		t.Fatalf("ss.client.EmptyCall(_, _) = _, %v (msg %q); want _, err with msg %q", err, status.Convert(err).Message(), wantMsg)
+	_, err := ss.client.EmptyCall(ctx, &testpb.Empty{})
+	st := status.Convert(err)
+	if st.Message() != wantMsg {
+		t.Fatalf("ss.client.EmptyCall(_, _) = _, %v (msg %q); want _, err with msg %q", err, st.Message(), wantMsg)
+	}
+	if len(st.Details()) != 0 {
+		// Details should be dropped on the server side.
+		t.Fatalf("RPC status contain details: %v, want no details", st.Details())
 	}
 }
 

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -6143,6 +6143,65 @@ func TestServeExitsWhenListenerClosed(t *testing.T) {
 	}
 }
 
+// Service handler returns status with invalid utf8 message.
+func TestStatusInvalidUTF8Message(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	var (
+		origMsg = string([]byte{0xff, 0xfe, 0xfd})
+		wantMsg = "���"
+	)
+
+	ss := &stubServer{
+		emptyCall: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			return nil, status.Errorf(codes.Internal, origMsg)
+		},
+	}
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	if _, err := ss.client.EmptyCall(ctx, &testpb.Empty{}); status.Convert(err).Message() != wantMsg {
+		t.Fatalf("ss.client.EmptyCall(_, _) = _, %v (msg %q); want _, err with msg %q", err, status.Convert(err).Message(), wantMsg)
+	}
+}
+
+// Service handler returns status with invalid utf8 details.
+func TestStatusInvalidUTF8Details(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	var (
+		origMsg = string([]byte{0xff, 0xfe, 0xfd})
+		wantMsg = "���"
+	)
+
+	ss := &stubServer{
+		emptyCall: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			st := status.New(codes.Internal, origMsg)
+			st, err := st.WithDetails(&testpb.Empty{})
+			if err != nil {
+				return nil, err
+			}
+			return nil, st.Err()
+		},
+	}
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	if _, err := ss.client.EmptyCall(ctx, &testpb.Empty{}); status.Convert(err).Message() != wantMsg {
+		t.Fatalf("ss.client.EmptyCall(_, _) = _, %v (msg %q); want _, err with msg %q", err, status.Convert(err).Message(), wantMsg)
+	}
+}
+
 func TestClientDoesntDeadlockWhileWritingErrornousLargeMessages(t *testing.T) {
 	defer leakcheck.Check(t)
 	for _, e := range listTestEnv() {

--- a/transport/http2_server.go
+++ b/transport/http2_server.go
@@ -38,6 +38,7 @@ import (
 	"google.golang.org/grpc/channelz"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/grpclog"
 	"google.golang.org/grpc/internal/grpcrand"
 	"google.golang.org/grpc/keepalive"
 	"google.golang.org/grpc/metadata"
@@ -768,10 +769,10 @@ func (t *http2Server) WriteStatus(s *Stream, st *status.Status) error {
 		stBytes, err := proto.Marshal(p)
 		if err != nil {
 			// TODO: return error instead, when callers are able to handle it.
-			panic(err)
+			grpclog.Errorf("transport: failed to marshal rpc status: %v, error: %v", p, err)
+		} else {
+			headerFields = append(headerFields, hpack.HeaderField{Name: "grpc-status-details-bin", Value: encodeBinHeader(stBytes)})
 		}
-
-		headerFields = append(headerFields, hpack.HeaderField{Name: "grpc-status-details-bin", Value: encodeBinHeader(stBytes)})
 	}
 
 	// Attach the trailer metadata.

--- a/transport/http_util.go
+++ b/transport/http_util.go
@@ -28,6 +28,7 @@ import (
 	"strconv"
 	"strings"
 	"time"
+	"unicode/utf8"
 
 	"github.com/golang/protobuf/proto"
 	"golang.org/x/net/http2"
@@ -442,11 +443,12 @@ const (
 )
 
 // encodeGrpcMessage is used to encode status code in header field
-// "grpc-message".
-// It checks to see if each individual byte in msg is an
-// allowable byte, and then either percent encoding or passing it through.
-// When percent encoding, the byte is converted into hexadecimal notation
-// with a '%' prepended.
+// "grpc-message". It does percent encoding and also replaces invalid utf-8
+// characters with Unicode replacement character.
+//
+// It checks to see if each individual byte in msg is an allowable byte, and
+// then either percent encoding or passing it through. When percent encoding,
+// the byte is converted into hexadecimal notation with a '%' prepended.
 func encodeGrpcMessage(msg string) string {
 	if msg == "" {
 		return ""
@@ -463,14 +465,16 @@ func encodeGrpcMessage(msg string) string {
 
 func encodeGrpcMessageUnchecked(msg string) string {
 	var buf bytes.Buffer
-	lenMsg := len(msg)
-	for i := 0; i < lenMsg; i++ {
-		c := msg[i]
-		if c >= spaceByte && c < tildaByte && c != percentByte {
-			buf.WriteByte(c)
-		} else {
-			buf.WriteString(fmt.Sprintf("%%%02X", c))
+	for len(msg) > 0 {
+		r, size := utf8.DecodeRuneInString(msg)
+		for _, b := range []byte(string(r)) {
+			if r >= spaceByte && r < tildaByte && r != percentByte {
+				buf.WriteByte(b)
+			} else {
+				buf.WriteString(fmt.Sprintf("%%%02X", b))
+			}
 		}
+		msg = msg[size:]
 	}
 	return buf.String()
 }

--- a/transport/http_util.go
+++ b/transport/http_util.go
@@ -468,9 +468,13 @@ func encodeGrpcMessageUnchecked(msg string) string {
 	for len(msg) > 0 {
 		r, size := utf8.DecodeRuneInString(msg)
 		for _, b := range []byte(string(r)) {
-			if r >= spaceByte && r < tildaByte && r != percentByte {
+			if size == 1 && b >= spaceByte && b < tildaByte && b != percentByte {
 				buf.WriteByte(b)
 			} else {
+				// If size > 1, r could be utf8.RuneError, and it is 2-bytes
+				// long. We need the for loop, otherwise
+				// fmt.Sprintf("%%%02X", utf8.RuneError) gives "%FFFD" instead
+				// of "%FF%FD".
 				buf.WriteString(fmt.Sprintf("%%%02X", b))
 			}
 		}

--- a/transport/http_util.go
+++ b/transport/http_util.go
@@ -468,13 +468,19 @@ func encodeGrpcMessageUnchecked(msg string) string {
 	for len(msg) > 0 {
 		r, size := utf8.DecodeRuneInString(msg)
 		for _, b := range []byte(string(r)) {
-			if size == 1 && b >= spaceByte && b < tildaByte && b != percentByte {
+			if size > 1 {
+				// If size > 1, r is not ascii. Always do percent encoding.
+				buf.WriteString(fmt.Sprintf("%%%02X", b))
+				continue
+			}
+
+			// The for loop is necessary even if size == 1. r could be
+			// utf8.RuneError.
+			//
+			// fmt.Sprintf("%%%02X", utf8.RuneError) gives "%FFFD".
+			if b >= spaceByte && b < tildaByte && b != percentByte {
 				buf.WriteByte(b)
 			} else {
-				// If size > 1, r could be utf8.RuneError, and it is 2-bytes
-				// long. We need the for loop, otherwise
-				// fmt.Sprintf("%%%02X", utf8.RuneError) gives "%FFFD" instead
-				// of "%FF%FD".
 				buf.WriteString(fmt.Sprintf("%%%02X", b))
 			}
 		}

--- a/transport/http_util_test.go
+++ b/transport/http_util_test.go
@@ -102,12 +102,14 @@ func TestEncodeGrpcMessage(t *testing.T) {
 	}{
 		{"", ""},
 		{"Hello", "Hello"},
-		{"my favorite character is \u0000", "my favorite character is %00"},
-		{"my favorite character is %", "my favorite character is %25"},
+		{"\u0000", "%00"},
+		{"%", "%25"},
+		{"系统", "%E7%B3%BB%E7%BB%9F"},
+		{string([]byte{0xff, 0xfe, 0xfd}), "%EF%BF%BD%EF%BF%BD%EF%BF%BD"},
 	} {
 		actual := encodeGrpcMessage(tt.input)
 		if tt.expected != actual {
-			t.Errorf("encodeGrpcMessage(%v) = %v, want %v", tt.input, actual, tt.expected)
+			t.Errorf("encodeGrpcMessage(%q) = %q, want %q", tt.input, actual, tt.expected)
 		}
 	}
 }
@@ -123,10 +125,36 @@ func TestDecodeGrpcMessage(t *testing.T) {
 		{"H%6", "H%6"},
 		{"%G0", "%G0"},
 		{"%E7%B3%BB%E7%BB%9F", "系统"},
+		{"%EF%BF%BD", "�"},
 	} {
 		actual := decodeGrpcMessage(tt.input)
 		if tt.expected != actual {
-			t.Errorf("dncodeGrpcMessage(%v) = %v, want %v", tt.input, actual, tt.expected)
+			t.Errorf("dncodeGrpcMessage(%q) = %q, want %q", tt.input, actual, tt.expected)
+		}
+	}
+}
+
+// Decode an encoded string should get the same thing back, except for invalid
+// utf8 chars.
+func TestDecodeEncodeGrpcMessage(t *testing.T) {
+	testCases := []struct {
+		orig string
+		want string
+	}{
+		{"", ""},
+		{"hello", "hello"},
+		{"h%6", "h%6"},
+		{"%G0", "%G0"},
+		{"系统", "系统"},
+		{"Hello, 世界", "Hello, 世界"},
+
+		{string([]byte{0xff, 0xfe, 0xfd}), "���"},
+		{string([]byte{0xff}) + "Hello" + string([]byte{0xfe}) + "世界" + string([]byte{0xfd}), "�Hello�世界�"},
+	}
+	for _, tC := range testCases {
+		got := decodeGrpcMessage(encodeGrpcMessage(tC.orig))
+		if got != tC.want {
+			t.Errorf("decodeGrpcMessage(encodeGrpcMessage(%q)) = %q, want %q", tC.orig, got, tC.want)
 		}
 	}
 }

